shadowsocks/udp.go

```diff
@@ -151,12 +146,17 @@ func (s *udpService) Start() {
 
 			payload := buf[len(tgtAddr):]
 
-			targetConn := nm.Get(clientAddr.String())
+			targetConn, clientLocation := nm.Get(clientAddr.String())
 			if targetConn == nil {
 				targetConn, err = net.ListenPacket("udp", "")
 				if err != nil {
 					return onet.NewConnectionError("ERR_CREATE_SOCKET", "Failed to create UDP socket", err)
 				}
+				clientLocation, locErr := s.m.GetLocation(clientAddr)
+				if locErr != nil {
+					logger.Warningf("Failed location lookup: %v", locErr)
+				}
+				logger.Debugf("Got location \"%v\" for IP %v", clientLocation, clientAddr.String())
```

Comment:
I think the `String()` call may be causing an allocation. Doesn't %v do the conversion fo ryou? Or maybe protect with an if. There's another Debugf call calling `.String()`.

Reply:
I agree.  I will work on this in a subsequent PR.